File: parse.c
/* parse.c - parsing and parse tree construction */

#include "ntlang.h"

void parse_table_init(struct parse_table_st *pt) {
    pt->len = 0;
}

struct parse_node_st * parse_node_new(struct parse_table_st *pt) {
    struct parse_node_st *np;

    np = &(pt->table[pt->len]);
    pt->len += 1;

    return np;
}

void parse_error(const char *err) {
    printf("%s\n", err);
    exit(-1);
}

void parse_error_message(const char *err, const char *message) {
    printf("%s%s\n", err, message);
    exit(-1);
}

char *parse_oper_strings[] = {"PLUS", "MINUS", "MULT", "DIV", "SHIFT_RIGHT", "SHIFT_LEFT", "ARITH_SHIFT_RIGHT", "BIT_AND", "BIT_OR", "BIT_XOR", "BIT_NOT"};

/* We need to provide prototypes for the parsing functions because
 * we call them before they are defined.
 */
struct parse_node_st * parse_program(struct parse_table_st *pt, 
                                        struct scan_table_st *st);
struct parse_node_st * parse_expression(struct parse_table_st *pt, 
                                        struct scan_table_st *st);
struct parse_node_st * parse_operand(struct parse_table_st *pt, 
                                        struct scan_table_st *st);

/* We need a parsing function for each rule in the EBNF grammer */

struct parse_node_st * parse_program(struct parse_table_st *pt, 
                                        struct scan_table_st *st) {
    struct parse_node_st *np1;

    /* A program is a single expression followed by EOT */
    np1 = parse_expression(pt, st);

    if (!scan_table_accept(st, TK_EOT)) {
        parse_error("Expecting EOT");
    }

    return np1;
}

struct parse_node_st * parse_expression(struct parse_table_st *pt, 
                                        struct scan_table_st *st) {
    struct scan_token_st *tp;
    struct parse_node_st *np1, *np2;

    /* An expression must start with an operand. */
    np1 = parse_operand(pt, st);

    while (true) {
        tp = scan_table_get(st, 0);
        /* Check for valid operator */
        if (tp->id == TK_PLUS || tp->id == TK_MINUS || tp->id == TK_MULT || tp->id == TK_DIV || tp->id == TK_SHIFT_RIGHT || tp->id == TK_SHIFT_LEFT || tp->id == TK_ARITH_SHIFT_RIGHT || tp->id == TK_BIT_AND || tp->id == TK_BIT_OR || tp->id == TK_BIT_XOR) {
            scan_table_accept(st, TK_ANY);
            np2 = parse_node_new(pt);
            np2->type = EX_OPER2;
            np2->oper2.oper = tp->id - TK_PLUS;
            np2->oper2.left = np1;
            /* parse second operand */
            np2->oper2.right = parse_operand(pt, st);
            np1 = np2;
        } else {
            break;
        }
    }
    return np1;
}

struct parse_node_st * parse_operand(struct parse_table_st *pt,
                                     struct scan_table_st *st) {
    struct scan_token_st *tp;
    struct parse_node_st *np1;

    if (scan_table_accept(st, TK_INTLIT)) {
            tp = scan_table_get(st, -1);
            np1 = parse_node_new(pt);
            np1->type = EX_LITERAL;
            np1->literal.value = 0;
            for (int i = 0; tp->value[i] != '\0'; i++) {
                if (tp->value[i] < '0' || tp->value[i] > '9') {
                    parse_error("Invalid integer literal");
                }
                np1->literal.value = np1->literal.value * 10 + (tp->value[i] - '0');
            }
        }else if (scan_table_accept(st, TK_HEXLIT)) {
             tp = scan_table_get(st, -1);
             np1 = parse_node_new(pt);
             np1->type = EX_LITERAL;
             np1->literal.value = 0;
             for (int i = 0; tp->value[i] != '\0'; i++) {
                 if (i >= 0) {
                     if (tp->value[i] >= '0' && tp->value[i] <= '9') {
                         // Check for overflow before performing addition
                         if (np1->literal.value <= (UINT32_MAX - (tp->value[i] - '0')) / 16) {
                             np1->literal.value = np1->literal.value * 16 + (tp->value[i] - '0');
                         } else {
                             parse_error_message("overflows uint32_t: ", tp->value);
                         }
                     } else if (tp->value[i] >= 'A' && tp->value[i] <= 'F') {
                         // Check for overflow before performing addition
                         if (np1->literal.value <= (UINT32_MAX - (tp->value[i] - 'A' + 10)) / 16) {
                             np1->literal.value = np1->literal.value * 16 + (tp->value[i] - 'A' + 10);
                         } else {
                             parse_error_message("overflows uint32_t: ", tp->value);
                         }
                     } else if (tp->value[i] >= 'a' && tp->value[i] <= 'f') {
                         // Check for overflow before performing addition
                         if (np1->literal.value <= (UINT32_MAX - (tp->value[i] - 'a' + 10)) / 16) {
                             np1->literal.value = np1->literal.value * 16 + (tp->value[i] - 'a' + 10);
                         } else {
                             parse_error_message("overflows uint32_t: ", tp->value);
                         }
                     } else {
                         parse_error("Invalid hexadecimal literal");
                     }
                 }
             }
         } else if (scan_table_accept(st, TK_BINLIT)) {
            tp = scan_table_get(st, -1);
            np1 = parse_node_new(pt);
            np1->type = EX_LITERAL;
            np1->literal.value = 0;
            for (int i = 0; tp->value[i] != '\0'; i++) {
                if (i >= 0) {
                    if (tp->value[i] == '0' || tp->value[i] == '1') {
                        np1->literal.value = np1->literal.value * 2 + (tp->value[i] - '0');
                    } else {
                        parse_error("Invalid binary literal");
                    }
                }
            }
        } else if (scan_table_accept(st, TK_MINUS)) {
            np1 = parse_node_new(pt);
            np1->type = EX_OPER1;
            np1->oper1.oper = OP_MINUS;
            np1->oper1.operand = parse_operand(pt, st);
        } else if (scan_table_accept(st, TK_BIT_NOT)) {
            np1 = parse_node_new(pt);
            np1->type = EX_OPER1;
            np1->oper1.oper = OP_BIT_NOT;
            np1->oper1.operand = parse_operand(pt, st);
        } else if (scan_table_accept(st, TK_LPAREN)) {
            np1 = parse_expression(pt, st);
            if (!scan_table_accept(st, TK_RPAREN)) {
                parse_error("Expecting ')'");
            }
        } else {
            parse_error("Bad operand");
        }
    return np1;
}

void parse_tree_print_indent(int level) {
    level *= 2;
    for (int i = 0; i < level; i++) {
        printf(".");
    }
}

void parse_tree_print_expr(struct parse_node_st *np, int level) {
    parse_tree_print_indent(level);
    printf("EXPR ");

    if (np->type == EX_LITERAL) {
        printf("LITERAL %d\n", np->literal.value);
    } else if (np->type == EX_OPER1) {
        printf("OPER1 %s\n", parse_oper_strings[np->oper1.oper]);
        parse_tree_print_expr(np->oper1.operand, level+1);
    } else if (np->type == EX_OPER2) {
        printf("OPER2 %s\n", parse_oper_strings[np->oper2.oper]);
        parse_tree_print_expr(np->oper2.left, level+1);
        parse_tree_print_expr(np->oper2.right, level+1);
    }
}

void parse_tree_print(struct parse_node_st *np) {
    parse_tree_print_expr(np, 0);    
}

File: src/parse.rs
use std::process;
use scan::{ScanTableSt, ScanToken};

#[derive(Debug, Clone)]
pub enum ParseOperator {
    Plus,
    Minus,
    Mult,
    Div,
    ShiftRight,
    ShiftLeft,
    ArithShiftRight,
    BitAnd,
    BitOr,
    BitXor,
    BitNot,
}

#[derive(Debug,Clone)]
pub enum ParseNodeType{
    Literal,
    Oper1,
    Oper2,
    None
}
#[derive(Debug,Clone)]
pub struct ParseNode {
    pub(crate) type_: ParseNodeType,
    pub(crate) value: i32,
    pub(crate) oper : ParseOperator,
    pub(crate) left: Option<Box<ParseNode>>,
    pub(crate) right: Option<Box<ParseNode>>,
}
impl ParseNode {
    fn new(value: i32) -> Self {
        ParseNode {
            type_: ParseNodeType::Literal,
            value,
            oper: ParseOperator::Plus,
            left: None,
            right: None,
        }
    }


}

pub struct ParseTableSt {
    pub table: Vec<ParseNode>, // Adjust the size as needed
    pub len: usize,
}

impl ParseTableSt {
    pub fn new() -> Self {
        ParseTableSt {
            table: Vec::new(),
            len: 0
        }
    }

    pub fn parse_node_new(&mut self) -> ParseNode {
        let node = ParseNode {
            type_: ParseNodeType::None,
            value: 0,
            oper: ParseOperator::Plus,
            left: None,
            right: None,
        };
        self.table.push(node.clone());
        self.len += 1;
        node
    }

    pub fn parse_program(&mut self, st: &mut ScanTableSt) -> Option<ParseNode> {
        let np1 = self.parse_expression(st)?;

        if !st.accept(ScanToken::EOT) {
            parse_error("Expecting EOT");
        }

        Some(np1)
    }

    pub fn parse_expression(&mut self, st: &mut ScanTableSt) -> Option<ParseNode> {
        let mut np1 = self.parse_operand(st)?;

        loop {
            let token = st.get(0);
            let is_operator = matches!(
                token.unwrap().id,
                ScanToken::Plus
                    | ScanToken::Minus
                    | ScanToken::Mult
                    | ScanToken::Div
                    | ScanToken::ShiftRight
                    | ScanToken::ShiftLeft
                    | ScanToken::ArithShiftRight
                    | ScanToken::BitAnd
                    | ScanToken::BitOr
                    | ScanToken::BitXor
            );

            if is_operator {
                let operator = match token.unwrap().id {
                    ScanToken::Plus => Some(ParseOperator::Plus),
                    ScanToken::Minus => Some(ParseOperator::Minus),
                    ScanToken::Mult => Some(ParseOperator::Mult),
                    ScanToken::Div => Some(ParseOperator::Div),
                    ScanToken::ShiftRight => Some(ParseOperator::ShiftRight),
                    ScanToken::ShiftLeft => Some(ParseOperator::ShiftLeft),
                    ScanToken::ArithShiftRight => Some(ParseOperator::ArithShiftRight),
                    ScanToken::BitAnd => Some(ParseOperator::BitAnd),
                    ScanToken::BitOr => Some(ParseOperator::BitOr),
                    ScanToken::BitXor => Some(ParseOperator::BitXor),
                    _ => unreachable!(),
                };
                st.accept(ScanToken::Any); // Consume the operator token
                let mut np2 = self.parse_node_new();

                np2.type_= ParseNodeType::Oper2;
                np2.oper = operator.unwrap();
                np2.left = Some(Box::new(np1));
                np2.right = Some(Box::new(self.parse_operand(st)?));
                np1 = np2;
            } else {
                break;
            }
        }
        Some(np1)
    }



    fn get_operator(&self, token: ScanToken) -> Option<ParseOperator> {
        match token {
            ScanToken::Plus => Some(ParseOperator::Plus),
            ScanToken::Minus => Some(ParseOperator::Minus),
            ScanToken::Mult => Some(ParseOperator::Mult),
            ScanToken::Div => Some(ParseOperator::Div),
            ScanToken::ShiftRight => Some(ParseOperator::ShiftRight),
            ScanToken::ShiftLeft => Some(ParseOperator::ShiftLeft),
            ScanToken::ArithShiftRight => Some(ParseOperator::ArithShiftRight),
            ScanToken::BitAnd => Some(ParseOperator::BitAnd),
            ScanToken::BitOr => Some(ParseOperator::BitOr),
            ScanToken::BitXor => Some(ParseOperator::BitXor),
            _ => None,
        }
    }

    pub fn parse_operand(&mut self, st: &mut ScanTableSt) -> Option<ParseNode> {
        if st.accept(ScanToken::IntLit) {
            Some(self.parse_literal_value(st, 10))
        } else if st.accept(ScanToken::HexLit) {
            Some(self.parse_literal_value(st, 16))
        } else if st.accept(ScanToken::BinLit) {
            Some(self.parse_literal_value(st, 2))
        } else if st.accept(ScanToken::Minus) {
            let mut np1 = self.parse_node_new();

            np1.type_= ParseNodeType::Oper1;
            np1.oper = self.get_operator(ScanToken::Minus).unwrap();
            np1.left = Some(Box::new(self.parse_operand(st)?));
            Some(np1)
        } else if st.accept(ScanToken::BitNot) {
            let mut np1 = self.parse_node_new();

            np1.type_= ParseNodeType::Oper1;
            np1.oper = self.get_operator(ScanToken::BitNot).unwrap();
            np1.left = Some(Box::new(self.parse_operand(st)?));
            Some(np1)
        } else if st.accept(ScanToken::LParen) {
            let np1 = self.parse_expression(st)?;
            if !st.accept(ScanToken::RParen) {
                parse_error("Expecting ')'");
            }
            Some(np1)
        } else {
            parse_error("Bad operand");
            None
        }
    }

    fn parse_literal_value(&mut self, st: &mut ScanTableSt, base: u32) -> ParseNode {
        let token = st.getLast(1).unwrap(); // Get the last scanned token
        let value = match base {
            10 => parse_literal_value_base_10(token.value.as_str()),
            16 => parse_literal_value_base_16(token.value.as_str()),
            2 => parse_literal_value_base_2(token.value.as_str()),
            _ => panic!("Unsupported base"),
        };
        let mut np1 = self.parse_node_new();

        np1.type_= ParseNodeType::Literal;
        np1.value = value as i32;
        np1
    }
}

fn parse_literal_value_base_10(value: &str) -> u32 {
    value.parse().unwrap() // Parse the string as u32
}

fn parse_literal_value_base_16(value: &str) -> u32 {
    u32::from_str_radix(value.trim_start_matches("0x"), 16).unwrap() // Parse hexadecimal
}

fn parse_literal_value_base_2(value: &str) -> u32 {
    u32::from_str_radix(value.trim_start_matches("0b"), 2).unwrap() // Parse binary
}

fn parse_error(err: &str) {
    println!("{}", err);
    process::exit(-1);
}

pub fn print_parse_tree(node: &ParseNode) {
    match node.type_ {
        ParseNodeType::Literal => println!("Literal: {}", node.value),
        ParseNodeType::Oper1 => {
            println!("Operator: {:?}", node.oper);
            if let Some(ref left) = node.left {
                print_parse_tree(left);
            }
        }
        ParseNodeType::Oper2 => {
            println!("Operator: {:?}", node.oper);
            if let Some(ref left) = node.left {
                print_parse_tree(left);
            }
            if let Some(ref right) = node.right {
                print_parse_tree(right);
            }
        }
        ParseNodeType::None => println!("Empty Node"),
    }
}
